oximeter/producer/src/lib.rs

```diff
@@ -64,56 +85,64 @@ pub struct Config {
     /// The information for contacting this server, and collecting its metrics.
     pub server_info: ProducerEndpoint,
     /// The address at which we attempt to register as a producer.
-    pub registration_address: SocketAddr,
+    ///
+    /// If the address is not provided, the address of Nexus will be resolved
+    /// using internal DNS, based on the local address of the server being
+    /// configured.
+    pub registration_address: Option<SocketAddr>,
     /// Configuration for starting the Dropshot server used to produce metrics.
     pub dropshot: ConfigDropshot,
     /// The logging configuration or actual logger used to emit logs.
     pub log: LogConfig,
 }
 
 /// A Dropshot server used to expose metrics to be collected over the network.
-///
-/// This is a "batteries-included" HTTP server, meant to be used in applications that don't
-/// otherwise run a server. The standalone functions [`register`] and [`collect`] can be used as
-/// part of an existing Dropshot server's API.
 pub struct Server {
     registry: ProducerRegistry,
+    registration_task: tokio::task::JoinHandle<()>,
     server: HttpServer<ProducerRegistry>,
 }
 
 impl Server {
     /// Start a new metric server, registering it with the chosen endpoint, and listening for
     /// requests on the associated address and route.
-    pub async fn start(config: &Config) -> Result<Self, Error> {
+    ///
+    /// Note that the producer server is registered with Nexus in a background
+    /// task. That task also periodically re-registers with Nexus to ensure that
+    /// data continues to be collected.
+    pub fn start(config: &Config) -> Result<Self, Error> {
         Self::with_registry(
             ProducerRegistry::with_id(config.server_info.id),
             &config,
         )
-        .await
     }
 
     /// Create a new metric producer server, with an existing registry.
-    pub async fn with_registry(
+    ///
+    /// Note that the producer server is registered with Nexus in a background
+    /// task.
+    pub fn with_registry(
         registry: ProducerRegistry,
         config: &Config,
     ) -> Result<Self, Error> {
         Self::new_impl(
             registry,
             config.server_info.clone(),
-            &config.registration_address,
+            config.registration_address.as_ref(),
             &config.dropshot,
             &config.log,
         )
-        .await
     }
 
     /// Serve requests for metrics.
     pub async fn serve_forever(self) -> Result<(), Error> {
+        self.registration_task.abort();
```

Comment:
I think that's right, and not what I intended! I think that slipped through because we don't actually need to call `serve_forever()` to start serving requests. That happens when we start the Dropshot server itself. In any case, I'll remove this, thanks for catching.

Reply:
Moved in 2c0afef2e